group.c

```diff
@@ -141,48 +165,77 @@ static int group_remove_regex(struct Group *g, const char *s)
   return mutt_regexlist_remove(&g->rs, s);
 }
 
-void mutt_group_context_add_addrlist(struct GroupContext *ctx, struct Address *a)
+void mutt_group_context_add_addrlist(struct GroupContextHead *head, struct Address *a)
 {
-  for (; ctx; ctx = ctx->next)
-    group_add_addrlist(ctx->g, a);
+  struct GroupContext *np = NULL;
+  STAILQ_FOREACH(np, head, entries)
+  {
+    group_add_addrlist(np->g, a);
+  }
 }
 
-int mutt_group_context_remove_addrlist(struct GroupContext *ctx, struct Address *a)
+int mutt_group_context_remove_addrlist(struct GroupContextHead *head, struct Address *a)
 {
   int rc = 0;
+  struct GroupContext *np = NULL, *tmp = NULL;
 
-  for (; (!rc) && ctx; ctx = ctx->next)
+  if (!rc)
   {
-    rc = group_remove_addrlist(ctx->g, a);
-    if (empty_group(ctx->g))
-      group_remove(ctx->g);
+    STAILQ_FOREACH_SAFE(np, head, entries, tmp)
+    {
+      if (!rc)
+      {
+        rc = group_remove_addrlist(np->g, a);
+        if (empty_group(np->g))
+          group_remove(np->g);
+      }
+      STAILQ_REMOVE(head, np, GroupContext, entries);
```

Comment:
The original function doesn't seem to be touching the elements of the list - `ctx` remains linked as it was before.